### PR TITLE
Fine tune diff highlighting

### PR DESCRIPTION
**Pull request by wxiaoguang**
Chroma seems extremely slow when highlighting large files, it might take dozens or hundreds of milliseconds.
When fully highlighting a diff with a lot of large files, it would take many seconds or even dozens of seconds.
So, don't highlight the entire file if it's too large, or highlighting takes too long.

Optimize `recoverOneDiff` a little with a improved test case and fix one more edge case.

### CODE REVIEW

**Comment by silverwind** [comment]
FYI, according to Google AI, GitHub applies these limits on syntax highlighting:

- **File Size Limit:** Syntax highlighting is disabled for files larger than **512 KB**.
- **Line Length:** Individual lines exceeding **1,000 characters** are often truncated or displayed without highlighting.
- **Diff Rendering:** In Pull Requests, syntax highlighting is limited to files smaller than **300 KB** and total diffs under **1 MB**.
- **Execution Time:** The highlighting engine will time out if parsing takes longer than **several seconds**.

**Comment by wxiaoguang** [comment]
> FYI, according to Google AI, GitHub applies these limits on syntax highlighting:
> 
>     * **File Size Limit:** Syntax highlighting is disabled for files larger than **512 KB**.
> 
>     * **Line Length:** Individual lines exceeding **1,000 characters** are often truncated or displayed without highlighting.
> 
>     * **Diff Rendering:** In Pull Requests, syntax highlighting is limited to files smaller than **300 KB** and total diffs under **1 MB**.
> 
>     * **Execution Time:** The highlighting engine will time out if parsing takes longer than **several seconds**.

We also have similar approaches.

File: services/gitdiff/gitdiff.go [hunk 1]
```diff
@@ -123,8 +124,13 @@ type DiffHTMLOperation struct {
 // BlobExcerptChunkSize represent max lines of excerpt
 const BlobExcerptChunkSize = 20
 
-// MaxDiffHighlightEntireFileSize is the maximum file size that will be highlighted with "entire file diff"
-const MaxDiffHighlightEntireFileSize = 1 * 1024 * 1024
+// Chroma seems extremely slow when highlighting large files, it might take dozens or hundreds of milliseconds.
+// When fully highlighting a diff with a lot of large files, it would take many seconds or even dozens of seconds.
+// So, don't highlight the entire file if it's too large, or highlighting takes too long.
+const (
+	MaxFullFileHighlightSizeLimit = 256 * 1024
```
Comment (silverwind) [inline services/gitdiff/gitdiff.go:132]: Use 512kB like GitHub or too much? Also I think it needs a settings variable so users can configure.

File: services/gitdiff/gitdiff.go [hunk 2]
```diff
@@ -123,8 +124,13 @@ type DiffHTMLOperation struct {
...
+	MaxFullFileHighlightTimeLimit = 2 * time.Second
```
Comment (silverwind) [inline services/gitdiff/gitdiff.go:133]: Also should have a settings variable imho.

**Review by silverwind — APPROVED** [review]

File: services/gitdiff/gitdiff.go (hunk 1 above)
Comment (wxiaoguang) [reply services/gitdiff/gitdiff.go:132]: Not need.

Even no "full file highlighting" , it will falls back to line-by-line highlighting.

256KB is pretty large for Chroma according to my test.

Maybe we need to write our owner highlighter if the performance needs to be improved.

File: services/gitdiff/gitdiff.go (hunk 2 above)
Comment (wxiaoguang) [reply services/gitdiff/gitdiff.go:133]: Not need.

Even no "full file highlighting" , it will falls back to line-by-line highlighting.

2 seconds is pretty long according to my test.

Maybe we need to write our owner highlighter if the performance needs to be improved.

File: services/gitdiff/gitdiff.go (hunk 1 above)
Comment (silverwind) [reply services/gitdiff/gitdiff.go:132]: But still, I think such things should be user-configurable. Thought, users shouldn't have to care about the difference between "full file" vs "line based" highlighting, e.g. should be simple variables:

- MAX_SYNTAX_HIGHLIGHTING_FILE_SIZE
- MAX_SYNTAX_HIGHLIGHTING_LINE_LENGTH
- MAX_SYNTAX_HIGHLIGHTING_TIME

etc

Comment (wxiaoguang) [reply services/gitdiff/gitdiff.go:132]: I don't see who really needs it.

If anyone needs it, there is always a chance to add them in the future.

https://en.wikipedia.org/wiki/Occam%27s_razor

Comment (silverwind) [reply services/gitdiff/gitdiff.go:132]: Well, stuff like "time" is always inprecise depending on CPU etc. 2s can be a lot on a beefy server but not much on a Rasperry Pi.

Comment (wxiaoguang) [reply services/gitdiff/gitdiff.go:132]: > But still, I think such things should be user-configurable. Thought, users shouldn't have to care about the difference between "full file" vs "line based" highlighting, e.g. should be simple variables:
> 
>     * MAX_SYNTAX_HIGHLIGHT_FILE_SIZE
> 
>     * MAX_SYNTAX_HIGHLIGHT_LINE_LENGTH
> 
>     * MAX_SYNTAX_HIGHLIGHT_TIME
> 
> 
> etc

By the way, the details are also complicated.

You have single file highlighting (view file), multiple file highlighting (diff view)

If you want to customize everything, you need M * N config options.

I really don't see why it's right to bloat the config system. We should provide the good-enough default behaviors to make users easy to use out-of-box.

Comment (wxiaoguang) [reply services/gitdiff/gitdiff.go:132]: > Well, stuff like "time" is always inprecise depending on CPU etc. 2s can be a lot on a beefy server but not much on a Rasperry Pi.

It is not decided by CPU, it is decided by user experience.

Why you'd like to open a page on Rasperry Pi which takes more than 2 or 5 seconds?

Comment (wxiaoguang) [reply services/gitdiff/gitdiff.go:132]: > 2s can be a lot on a beefy server

If you think so, either you should upgrade the server, or you can also disable the highlighting (there is already an option). Other rendering logic can still take up to several seconds even if there is no highlight.

Comment (silverwind) [reply services/gitdiff/gitdiff.go:132]: You can't know what the user priorizes more: render speed vs. accurate highlighting.

Comment (wxiaoguang) [reply services/gitdiff/gitdiff.go:132]: > You can't know what the user priorizes more: render speed vs. accurate highlighting.

Can you just don't guess and imagine? Show me a real case why you need it or any end user needs it.

Have you fully understood the problem?  https://github.com/go-gitea/gitea/pull/36592#discussion_r2796814057

Does my code block anything or make anything worse?  https://github.com/go-gitea/gitea/pull/36592#discussion_r2796807139

> > If anyone needs it, there is always a chance to add them in the future.

Comment (silverwind) [reply services/gitdiff/gitdiff.go:132]: Let's stop this discussion.

**Review by lunny — APPROVED** [review]
